src/table/src/metadata.rs

```diff
@@ -817,17 +817,33 @@ impl TableMeta {
             );
         }
         // Collect columns after changed.
-        let columns: Vec<_> = table_schema
-            .column_schemas()
-            .iter()
-            .cloned()
-            .map(|mut column| {
-                if let Some(change_column) = modify_column_types.get(&column.name) {
-                    column.data_type = change_column.target_type.clone();
-                }
-                column
-            })
-            .collect();
+        let mut columns: Vec<_> = vec![];
+        for mut column in table_schema.column_schemas().iter().cloned() {
+            if let Some(change_column) = modify_column_types.get(&column.name) {
+                column.data_type = change_column.target_type.clone();
+                let new_default = if let Some(default_value) = column.default_constraint() {
+                    Some(
+                        default_value
+                            .cast_to_datatype(&change_column.target_type)
+                            .with_context(|_| error::CastDefaultValueSnafu {
+                                reason: format!(
+                                    "Failed to cast default value from {:?} to type {:?}",
+                                    default_value, &change_column.target_type
+                                ),
+                            })?,
+                    )
+                } else {
+                    None
+                };
+                column = column
+                    .clone()
+                    .with_default_constraint(new_default.clone())
```

Comment:
Seems these two clones are not needed:

```suggestion
                    .with_default_constraint(new_default)
```

Reply:
> Seems these two clones are not needed:

was used in the error msg below, also not too expensive to clone anyway